### PR TITLE
Feature/targeted spectra extractor store spectra for pr

### PR DESCRIPTION
**Pull request by bertrandboudaud**
# Description

Add some basic support for integrating spectral information when conducting data dependent acquisition (DDA) metabolomics experiments
  * searchSpectrum
  * mergeFeatures
  * annotateSpectra
  * storeSpectraTraML
  * storeSpectraMSP

# Checklist:
- [X] Make sure that you are listed in the AUTHORS file
- [X] Add relevant changes and new features to the CHANGELOG file
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] New and existing unit tests pass locally with my changes
- [X] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.

### CODE REVIEW

**Comment by oliveralka** [comment]
rebuild jenkins

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h [hunk 1]
```diff
@@ -415,11 +443,45 @@ namespace OpenMS
...
 protected:
     /// Overridden function from DefaultParamHandler to keep members up to date, when a parameter is changed
     void updateMembers_() override;
 
-private:
+    /// Deisotope MS2 spectra
+    void deisotopeMS2Spectra_(MSExperiment& experiment) const;
+
+    /// Remove peaks form MS2 which are at a higher mz than the precursor + 10 ppm
+    void removeMS2SpectraPeaks_(MSExperiment& experiment) const;
```
Comment (oliveralka) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:479]: We have to keep in mind/comment, that this only works correctly for compounds with charge 1.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h [hunk 2]
```diff
@@ -207,6 +207,34 @@ namespace OpenMS
       std::vector<MSSpectrum>& annotated_spectra
     ) const;
 
+    /**
+      @brief Annotates the MS2 spectra with the likely MS1 feature that it was derived from
+
+      Add support for annotating based off of MS1 features resulting from SearchAccurateMass.
```
Comment (oliveralka) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:213]:
```suggestion
      Annotating based on MS1 feature results from AccurateMassSearch.
```

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h [hunk 3]
```diff
@@ -207,6 +207,34 @@ namespace OpenMS
...
+      In this case, the input will be e.g., const FeatureMap& ms1_features and the RTs and names (i.e., PeptideRef),
+      defined in the FeatureMap.
+
+      @param[in] spectra The spectra to filter
+      @param[in] ms1_features the MS1 features
```
Comment (oliveralka) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:218]:
```suggestion
      @param[in] ms1_features The MS1 features
```

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h [hunk 4]
```diff
@@ -207,6 +207,34 @@ namespace OpenMS
...
+      @param[out] ms1_features the MS2 features
```
Comment (oliveralka) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:219]:
```suggestion
      @param[out] ms1_features The MS2 features
```

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h [hunk 5]
```diff
@@ -207,6 +207,34 @@ namespace OpenMS
...
+        FeatureMap& ms2_features,
+        std::vector<MSSpectrum>& annotated_spectra) const;
+    
+    /**
+      @brief Search accurate masses and build add the the peptide hits as features/sub-features
+
+      @param[in] feat_map The feature map to search in
+      @param[in] feat_map_output The output feature map, with peptide identifaction as sub features
+    */
+    void searchSpectrum(
```
Comment (oliveralka) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:234]: I think I have not yet understood what this is used for - based on the description.

**Review by oliveralka — COMMENTED** [review]
I think that looks good so far. 
Thanks for the addition. 

I think we have be careful with the deisotoper parameters and with the removal of all peaks above the precursor (only valid for charge 1) @timosachsenberg What do you think?

File: src/pyOpenMS/pxds/TargetedSpectraExtractor.pxd [hunk 6]
```diff
@@ -32,6 +35,11 @@ cdef extern from "<OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h>" namesp
         void extractSpectra(MSExperiment&, TargetedExperiment&, libcpp_vector[ MSSpectrum ]&, FeatureMap&) nogil except +
         void extractSpectra(MSExperiment&, TargetedExperiment&, libcpp_vector[ MSSpectrum ]&) nogil except +
 
+        void storeSpectraTraML(const String&, FeatureMap&, FeatureMap&) nogil except +
+        void storeSpectraMSP(const String&, FeatureMap&, MSExperiment&) nogil except +
```
Comment (oliveralka) [inline src/pyOpenMS/pxds/TargetedSpectraExtractor.pxd:39]:
```suggestion
        void storeSpectraMSP(const String&, MSExperiment&) nogil except +
```

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h (hunk 1 above)
Comment (timosachsenberg) [reply src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:479]: yes the charge 1 precursor m/z as reference should be fine.
But if we have a charge 1 precursor and you want to keep the isotope pattern then
you need to choose a larger tolerance

**Comment by oliveralka** [comment]
rebuild jenkins

**Comment by bertrandboudaud** [comment]
@oliveralka Hi, i'm still having some compilation issues, however, i cannot spot the reasons clearly. i can see for ubuntu

``
   '!' represents an error and '*' a warning.

    ..................................................  Size: 49K

    .................*.....*..........................  Size: 99K

    ..........................!****************.******  Size: 149K

    **********************************.*************!*  Size: 199K

    ***************.**********************************  Size: 249K

    ********************!.............................  Size: 299K

    ..................................................  Size: 349K

    ..................................................  Size: 399K

    ..................................................  Size: 449K

    ........................... Size of output: 477K

Error(s) when building project
Starting class and or tool tests...
Unable to find executable: /home/jenkins/ws/openms/PR/bldtst/a9154670/build/src/tests/class_tests/bin/TargetedSpectraExtractor_test
``

However there is not much details. is there a way to get the details?
(same with apple build... event more opaque since I don't see any error in log)

thank you :)

**Comment by oliveralka** [comment]
Let me check!
Comment above refers to:
https://abibuilder.informatik.uni-tuebingen.de/jenkins/job/openms/job/PR/job/bldtst/1162/

**Comment by jpfeuffer** [comment]
@bertrandboudaud Thanks for the PR! 
Regarding test logs. I would always start on our dashboard where the test results get uploaded in a shortened but readable overview: http://cdash.openms.de/index.php?project=OpenMS (look for your PR number in the beginning of the builds in the "Continuous" section).
The Jenkins logs contain the nitty-gritty details.

A general question: Can you provide/describe some kind of "workflow/application" on how/when to use the new functionality? I am working on some Metabolomics applications and this sounds interesting. We could also discuss this on the GitHub discussions page or in the next OpenMS video call.

**Comment by bertrandboudaud** [comment]
Thank you both for your help with the log, I'm going to correct that for apple and ubuntu, thanks to your CI (it work on my side on my ubuntu and windows).
as for the details and a meeting, I will synchronize with Douglas when we can have that as he had led my work on that.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h (hunk 3 above)
Comment (oliveralka) [reply src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:218]: Can we keep that consistent in the header?

File: src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp [hunk 7]
```diff
@@ -146,6 +167,136 @@ namespace OpenMS
...
+  }
+
+  void TargetedSpectraExtractor::searchSpectrum(
+      OpenMS::FeatureMap& feat_map,
+      OpenMS::FeatureMap& feat_map_output) const
+  {
+    OpenMS::AccurateMassSearchEngine ams;
+    OpenMS::MzTab output;
+    ams.init();
+    ams.run(feat_map, output);
```
Comment (oliveralka) [inline src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:255]: When AMSE is initialized inside the `searchSpectrum()`, the database and adduct list can not be updated, or? Which means the algorithm, only works with the current HMDB (4.0) database. There is no way to use a custom database or to reduce the adduct space. 

Is that the desired, or should there be a possibility to change the default?

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h (hunk 1 above)
Comment (oliveralka) [reply src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:479]: I think adding a parameter is a good idea!

**Comment by oliveralka** [comment]
rebuild jenkins

**Comment by bertrandboudaud** [comment]
rebuild jenkins

**Comment by bertrandboudaud** [comment]
StopWatch_test failed for some reasons that does not seem to be related with the modifications of this PR. how can I relaunch jenkins ?

**Comment by bertrandboudaud** [comment]
Having hard time with StopWatch test (if i read well the logs) .. does not look to be related with the modification of this PR.

**Comment by bertrandboudaud** [comment]
rebuild jenkins

**Comment by timosachsenberg** [comment]
The StopWatch test seems to be unrelated so don't worry

**Comment by bertrandboudaud** [comment]
rebuild jenkins

**Comment by bertrandboudaud** [comment]
rebuild jenkins

**Comment by jpfeuffer** [comment]
Bildschirmfoto 2021-06-21 um 12 37 29

I think travis has Warn Unused = error

File: src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp [hunk 8]
```diff
@@ -146,6 +173,139 @@ namespace OpenMS
...
+        for (const auto& subordinate : feature.getSubordinates())
+        {
+          const auto& peptide_ref = subordinate.getMetaValue("PeptideRef");
+          const double target_mz = subordinate.getMZ();
+          const double target_rt = subordinate.getRT();
+          if (target_rt >= rt_left_lim && target_rt <= rt_right_lim)
+          {
+            std::cout << "annotateSpectra(): " << peptide_ref << "]";
+            std::cout << " (target_rt: " << target_rt << ") (target_mz: " << target_mz << ")" << std::endl
+                      << std::endl;
```
Comment (jpfeuffer) [inline src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:239]: Could you switch to one of the Logging mechanisms here?

File: src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp [hunk 9]
```diff
@@ -146,6 +173,139 @@ namespace OpenMS
...
+      {
+        OPENMS_LOG_WARN << "annotateSpectra(): No precursor MZ found. Setting spectrum_mz to 0." << std::endl;
+      }
+      const double spectrum_mz = precursors.empty() ? 0.0 : precursors.front().getMZ();
+      const double mz_tolerance = mz_unit_is_Da_ ? mz_tolerance_ : mz_tolerance_ / 1e6;
+
+      // When spectrum_mz is 0, the mz check on transitions is inhibited
+      const double mz_left_lim = spectrum_mz ? spectrum_mz - mz_tolerance : std::numeric_limits<double>::min();
+      const double mz_right_lim = spectrum_mz ? spectrum_mz + mz_tolerance : std::numeric_limits<double>::max();
+
```
Comment (jpfeuffer) [inline src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:227]: apparently unused?

File: src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp [hunk 10]
```diff
@@ -313,12 +473,14 @@ namespace OpenMS
       }
 
       double avgFWHM { 0 };
-      for (Size j = 0; j < picked_spectra[i].getFloatDataArrays()[0].size(); ++j)
+      if (picked_spectra[i].getFloatDataArrays().size())
```
Comment (jpfeuffer) [inline src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:476]: Can you use !empty here?

Comment (jpfeuffer) [reply src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:476]: And in the following instances where size != 0 is used to implicitly ask for !empty.

**Review by jpfeuffer — CHANGES_REQUESTED** [review]
Almost done.

**Comment by bertrandboudaud** [comment]
rebuild jenkins

**Review by jpfeuffer — APPROVED** [review]

**Comment by jpfeuffer** [comment]
remaining failures are unrelated

**Comment by bertrandboudaud** [comment]
> remaining failures are unrelated

Thanks for your help and review!

**Comment by bertrandboudaud** [comment]
@jpfeuffer @oliveralka can we merge this ? is there anything I should do more? (I can't merge myself, right?)

**Pull request merged**